### PR TITLE
EarlyCSE: refactor getOrCreateResult (NFC)

### PR DESCRIPTION
**Pull request by artagnon**

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp [hunk 1]
```diff
@@ -964,32 +965,45 @@ class EarlyCSE {
...
-      Value *V = SI->getValueOperand();
-      return V->getType() == ExpectedType ? V : nullptr;
+  Value *getOrCreateResult(Instruction *Inst, Type *ExpectedType) const {
+    if (!isa<IntrinsicInst, LoadInst, StoreInst>(Inst))
+      llvm_unreachable("Instruction not supported");
+
+    // The load or the store's first operand.
+    Value *V;
+    if (auto *II = dyn_cast<IntrinsicInst>(Inst)) {
+      if (isHandledNonTargetIntrinsic(II->getIntrinsicID()))
```
Comment (arsenm) [inline llvm/lib/Transforms/Scalar/EarlyCSE.cpp:975]: isHandledNonTargetIntrinsic is redundant with just switching over the handled IDs

File: llvm/test/CodeGen/PowerPC/big-endian-store-forward.ll [hunk 2]
```diff
@@ -1,12 +1,17 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
Comment (arsenm) [inline llvm/test/CodeGen/PowerPC/big-endian-store-forward.ll:1]: Should avoid converting tests to generated checks in functional commits

Comment (goldsteinn) [reply llvm/test/CodeGen/PowerPC/big-endian-store-forward.ll:1]: If you need to do so, pre-commit regeneration of the tests. Possibly under another PR.

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp [hunk 3]
```diff
@@ -964,32 +965,45 @@ class EarlyCSE {
...
-  Value *getOrCreateResult(Value *Inst, Type *ExpectedType) const {
-    // TODO: We could insert relevant casts on type mismatch here.
-    if (auto *LI = dyn_cast<LoadInst>(Inst))
-      return LI->getType() == ExpectedType ? LI : nullptr;
-    if (auto *SI = dyn_cast<StoreInst>(Inst)) {
-      Value *V = SI->getValueOperand();
-      return V->getType() == ExpectedType ? V : nullptr;
+  Value *getOrCreateResult(Instruction *Inst, Type *ExpectedType) const {
+    if (!isa<IntrinsicInst, LoadInst, StoreInst>(Inst))
+      llvm_unreachable("Instruction not supported");
```
Comment (goldsteinn) [inline llvm/lib/Transforms/Scalar/EarlyCSE.cpp:970]: I would just say "assert(isa<...>(Inst) && "Instruction not supported");`

File: llvm/test/CodeGen/PowerPC/big-endian-store-forward.ll (hunk 2 above)
Comment (artagnon) [reply llvm/test/CodeGen/PowerPC/big-endian-store-forward.ll:1]: > Based on #113338.

which contains the regen-with-UTC. Kindly look only at the second commit.

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp [hunk 4]
```diff
@@ -31,6 +31,7 @@
 #include "llvm/IR/Constants.h"
 #include "llvm/IR/Dominators.h"
 #include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
```
Comment (goldsteinn) [inline llvm/lib/Transforms/Scalar/EarlyCSE.cpp:34]: For some reason I thought we werent allowed to use builder in EarlyCSE, although thinking about the reason I don't have anythiny concrete.

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp [hunk 5]
```diff
@@ -964,32 +965,42 @@ class EarlyCSE {
...
-  Value *getOrCreateResult(Value *Inst, Type *ExpectedType) const {
-    // TODO: We could insert relevant casts on type mismatch here.
-    if (auto *LI = dyn_cast<LoadInst>(Inst))
-      return LI->getType() == ExpectedType ? LI : nullptr;
-    if (auto *SI = dyn_cast<StoreInst>(Inst)) {
-      Value *V = SI->getValueOperand();
-      return V->getType() == ExpectedType ? V : nullptr;
+  Value *getOrCreateResult(Instruction *Inst, Type *ExpectedType) const {
+    assert((isa<IntrinsicInst, LoadInst, StoreInst>(Inst)) &&
+           "Instruction not supported");
```
Comment (arsenm) [inline llvm/lib/Transforms/Scalar/EarlyCSE.cpp:970]: Redundant with the casts below

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp [hunk 6]
```diff
@@ -964,32 +965,39 @@ class EarlyCSE {
...
+    // First handle the case when no cast is required.
+    if (ActualType == ExpectedType)
+      return V;
+
+    // Try to create BitCast, SExt, or Trunc.
+    IRBuilder<> Builder(TheBB, std::next(Inst->getIterator()));
+    if (CastInst::castIsValid(Instruction::BitCast, V, ExpectedType))
+      return Builder.CreateBitCast(V, ExpectedType);
+    if (CastInst::castIsValid(Instruction::SExt, V, ExpectedType))
+      return Builder.CreateSExt(V, ExpectedType);
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/EarlyCSE.cpp:998]: This is very incorrect. If you have an i32 store and then an i64 load, the results of the i64 load will not be the sext of the i32 store. In EarlyCSE you don't have knowledge about how the high bits looks like.

Generating a bitcast here is fine, but let's please not produce value changing casts.

**Review by nikic — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp (hunk 6 above)
Comment (artagnon) [reply llvm/lib/Transforms/Scalar/EarlyCSE.cpp:998]: I had  my doubts about this part too; thanks for sanity-checking: will drop the sext and trunc.

**Review by nikic — CHANGES_REQUESTED** [review]
The idea here is reasonable, but I don't think you can create a bitcast in getOrCreateResult(). This function may be called in cases where we will not actually use the result, such as https://github.com/llvm/llvm-project/blob/c04719379e898a0159ac41999ea48c8e89b35160/llvm/lib/Transforms/Scalar/EarlyCSE.cpp#L1716. In this case we will be leaving behind a dead bitcast.

**Comment by artagnon** [comment]
> The idea here is reasonable, but I don't think you can create a bitcast in getOrCreateResult(). This function may be called in cases where we will not actually use the result, such as
> 
> https://github.com/llvm/llvm-project/blob/c04719379e898a0159ac41999ea48c8e89b35160/llvm/lib/Transforms/Scalar/EarlyCSE.cpp#L1716
> . In this case we will be leaving behind a dead bitcast.

Makes sense. I do think that the refactoring of getOrCreateResult is a good change, so I've made this patch an NFC doing that.

**Review by nikic — APPROVED** [review]
LG, but I think the TODO should be preserved. The general idea of supporting bitcasts makes sense, it's just not that straightforward to implement.

**Pull request merged**